### PR TITLE
Updated mantidworkbench to send stacktrace

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
**Description of work.**

This updates the error reporter in two ways
  * It removes the sending of recovery files as these were found to be not very useful.
  * It adds in the sending of a stacktrace in workbench in cases where one can be retrieved.

When this is merged we should update the [privacy policy](https://www.mantidproject.org/MantidProject:Privacy_policy#Usage_Data_recorded_in_Mantid) to reflect what is now sent.

**To test:**
  * To test this properly you will need to get a local version of the error report server with this pull request built. https://github.com/mantidproject/errorreports/pull/15
  * You will also need to change the errorreports url in Mantid.properties.template to http://localhost:8082
  * There are two cases which need to be tested. 
       * Testing the catching of non-crashing errors can be done by trying to cancel a mantid build via a keyboard interrupt. 
       * To test the other case you need to add `raise RuntimeError("Error Message") before line 859 in mainwindow.py . This will then provoke an excetion on startip.

  * Check that in both cases check that error-reports are received by the server with appropriate stack traces.

*There is no associated issue.*

*This does not require release notes* because this is a developer facing change

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DanielMurphy22** [comment]
Code works as expected, just the one build failure

**Review by DanielMurphy22 — DISMISSED** [review]

File: Framework/Kernel/inc/MantidKernel/ErrorReporter.h [hunk 1]
```diff
@@ -60,7 +59,7 @@ class MANTID_KERNEL_DLL ErrorReporter {
   /// Target url
   std::string m_url;
   /// recovery file path
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/ErrorReporter.h]: Can we either update the comments or remove them for all of these member variables? The names are good so I think the comments are not really required.

File: qt/applications/workbench/workbench/app/mainwindow.py [hunk 2]
```diff
@@ -864,6 +864,10 @@ def main():
         # about. Prints to stderr as we can't really count on anything
         # else
         traceback.print_exc(file=ORIGINAL_STDERR)
+        print_file_path = os.path.join(ConfigService.getAppDataDirectory(), 'workbench_stacktrace.txt')
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/app/mainwindow.py]: Couple of things:
* can we define the string `'workbench_stacktrace.txt'` using a named constant put at the top of the module with the others?
* we need to be careful with file handling as it's entirely possible to get an unexpected `OSError`. Can we wrap this in a `try...except` and also use `with open ...` so we don't need to worry about `.close()`

File: scripts/ErrorReporter/error_report_presenter.py [hunk 3]
```diff
@@ -10,20 +10,29 @@
...
         self.error_log = Logger("error")
         self._view = view
         self._exit_code = exit_code
         self._application = application
+        self._traceback = traceback
         self._view.set_report_callback(self.error_handler)
 
+        if not traceback:
+            traceback_file_path = os.path.join(ConfigService.getAppDataDirectory(), '{}_stacktrace.txt'.format(application))
+            if os.path.isfile(traceback_file_path):
```
Comment (martyngigg) [inline scripts/ErrorReporter/error_report_presenter.py]: Similarly we need to make sure we handle errors around the os calls. Can we use `with` for opening the file and I think `open` and `os.rename` need an `OSError` catcher for safety?

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: qt/applications/workbench/workbench/app/mainwindow.py [hunk 4]
```diff
@@ -864,10 +865,12 @@ def main():
         # about. Prints to stderr as we can't really count on anything
         # else
         traceback.print_exc(file=ORIGINAL_STDERR)
-        print_file_path = os.path.join(ConfigService.getAppDataDirectory(), 'workbench_stacktrace.txt')
-        print_file = open(print_file_path, 'w')
-        traceback.print_exc(file=print_file)
-        print_file.close()
+        try:
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/app/mainwindow.py]: :+1:

**Review by martyngigg — DISMISSED** [review]

**Comment by martyngigg** [comment]
@Mantid-Matthew There are a couple of tests failing now. I expect because we have removed the recovery info.

**Comment by martyngigg** [comment]
@Mantid-Matthew Would you have time to look at the failing tests here?

**Comment by martyngigg** [comment]
@Mantid-Matthew Is there any issue with the Windows tests?

**Comment by Mantid-Matthew** [comment]
No I was trying to build a windows package on another branch and lost track of which tab I was on.

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers Any chance we could get this merged to get it out to users? The stacktrace will be very useful for us figuring out what has happened.

The `Ubuntu Python 3` is now an orphaned status as we have removed that job since the default is now Python 3.

**Pull request merged**